apps/src/music/blockly/blocks/simple2.js

```diff
@@ -240,6 +249,120 @@ export const playSoundsSequential = {
       `
 };
 
+export const playSoundsRandom = {
+  definition: {
+    type: BlockTypes.PLAY_SOUNDS_RANDOM,
+    message0: 'play random',
+    args0: [],
+    message1: '%1',
+    args1: [
+      {
+        type: 'input_statement',
+        name: 'code'
+      }
+    ],
+    inputsInline: true,
+    previousStatement: null,
+    nextStatement: null,
+    style: 'flow_blocks',
+    tooltip: 'play sound randomly',
+    helpUrl: ''
+  },
+  generator: block => {
+    const resultArray = [];
+    let currentBlock = block.getInputTargetBlock('code');
+    while (currentBlock) {
+      const codeForBlock = getCodeForSingleBlock(currentBlock);
+      resultArray.push(codeForBlock);
+      currentBlock = currentBlock.getNextBlock();
+    }
+
+    const randomVar = Blockly.JavaScript.nameDB_.getDistinctName(
+      'random',
+      Blockly.Names.NameType.VARIABLE
+    );
+
+    let code = `var ${randomVar} = Math.floor(Math.random() * ${
+      resultArray.length
+    });`;
+
+    for (const [resultIndex, result] of resultArray.entries()) {
+      const lastMaybeSkipSoundVar = Blockly.JavaScript.nameDB_.getDistinctName(
+        '__lastMaybeSkipSound',
+        Blockly.Names.NameType.VARIABLE
+      );
+      const lastSkipSoundVar = Blockly.JavaScript.nameDB_.getDistinctName(
+        '__lastSkipSound',
+        Blockly.Names.NameType.VARIABLE
+      );
+      code += `
+        ${lastMaybeSkipSoundVar} = __maybeSkipSound;
+        ${lastSkipSoundVar} = __skipSound;
```

Comment:
Not sure if it would be much simpler, but I'm wondering if it may be easier to hoist this logic out of generator code and into some sort of `RandomSkipManager` analogous to `ProgramSequencer`? Thinking out loud, but 

```
RandomManager.newSkipContext(length); // starts a new random skip context and pushes a new frame onto the stack. At this point, RandomManager randomly chooses an index from the given length

// for each code result
${result}
RandomSkipManager.next(); // increments the index for the current skip context
...

RandomSkipManager.endSkipContext(); // ends random context, pops frame from stack and returns to previous context
```

and within the play generator, `RandomSkipManager.getSkipContext()` would return the current skip context -
```
{
   skippable: if RandomSkipManager currently has an active skip context,
   skipSound: if RandomSkipManager's current index != the random index it had picked when the skip context was created 
}
```

Reply:
Great call.  I had a feeling this would be a good idea, and it was certainly a non-trivial amount of work, but it was fun to do more unit-testing, and it makes the generated code a lot nicer to look at... so I did it!